src/symbolizer.cc

```diff
@@ -45,49 +59,38 @@ DDRes Symbolizer::symbolize_pprof(std::span<ElfAddress_t> elf_addrs,
                                   std::span<ddog_prof_Location> locations,
                                   unsigned &write_index,
                                   BlazeResultsWrapper &results) {
-  blaze_symbolizer *symbolizer = nullptr;
-  if (elf_addrs.size() != process_addrs.size()) {
+  if (elf_addrs.size() != process_addrs.size() || elf_addrs.empty() ||
+      elf_src.empty()) {
     LG_WRN("Error in provided addresses when symbolizing pprofs");
     return ddres_warn(DD_WHAT_PPROF); // or some other error handling
   }
-  if (elf_addrs.empty() || elf_src.empty()) {
-    return ddres_warn(DD_WHAT_PPROF); // or some other error handling
-  }
-  ddprof::HeterogeneousLookupStringMap<std::string> *demangled_names = nullptr;
-  bool retry_symbolization = true;
-  bool use_debug = true;
-  const blaze_result *blaze_res = nullptr;
-  while (retry_symbolization && !_disable_symbolization) {
-    const auto it = _symbolizer_map.find(file_id);
-    const char *resolved_src = elf_src.c_str();
-    // This is to avoid we change the path at every call (for different pids)
-    // The cache takes into account the first path given
-    if (it != _symbolizer_map.end()) {
-      resolved_src = it->second._elf_src.c_str();
-      symbolizer = it->second._symbolizer.get();
-      it->second._visited = true;
-      demangled_names = &(it->second._demangled_names);
-      retry_symbolization = false;
-      use_debug = it->second._use_debug;
-    } else {
-      auto pair = _symbolizer_map.emplace(
-          file_id,
-          BlazeSymbolizerWrapper(elf_src, inlined_functions, use_debug));
-      DDPROF_DCHECK_FATAL(pair.second, "Unable to insert symbolizer object");
-      symbolizer = pair.first->second._symbolizer.get();
-      demangled_names = &(pair.first->second._demangled_names);
-    }
-    // Initialize the src_elf structure
-    const blaze_symbolize_src_elf src_elf{
+
+  if (!_disable_symbolization) {
+    auto &symbolizer_wrapper = get_symbolizer(file_id, elf_src);
+
+    blaze_symbolize_src_elf src_elf{
         .type_size = sizeof(blaze_symbolize_src_elf),
-        .path = resolved_src,
-        .debug_syms = use_debug,
+        .path = symbolizer_wrapper.elf_src.c_str(),
+        .debug_syms = symbolizer_wrapper.use_debug,
         .reserved = {},
     };
 
     // Symbolize the addresses
-    blaze_res = blaze_symbolize_elf_virt_offsets(
-        symbolizer, &src_elf, elf_addrs.data(), elf_addrs.size());
+    const auto *blaze_res = blaze_symbolize_elf_virt_offsets(
+        symbolizer_wrapper.symbolizer.get(), &src_elf, elf_addrs.data(),
+        elf_addrs.size());
+    if (!blaze_res && symbolizer_wrapper.use_debug) {
```

Comment:
Do you want to change to a retry at every symbolization attempt ?

Reply:
Hum, it should only retry on the first symbolization attempt:
* on first attempt `symbolizer_wrapper.use_debug` is `true`, so symbolizer is called with debug symbols
* if it fails, `symbolizer_wrapper.use_debug` is set to false and symbolizer is called again without debug symbols
* on next attempt, `symbolizer_wrapper.use_debug` is false, so symbolizer is called without debug symbols and no other attempt is made is it fails

Reply:
That makes sense :+1: